### ISSUE TITLE
Question: How to call glibc function or third part library in wasm

### ISSUE THREAD

**Issue by mrpre** [wasmerio/wasmer#522]
I'm using `wasmer-c-api` and trying to embed wasmer into C application.  Everything is ok but if i using some glibc function like `malloc` or `free`, sample runs fail.  

source.c 
```
#include <stdio.h>
#include <malloc.h>
#include <string.h>
extern void print_str(char *ptr, int len);

int hello_wasm()
{
    char *str = "Hello, World!";
    print_str(str, 13);
    char *ptr = malloc(100);
    memcpy(ptr, "Goodbye, World!", 16);
    print_str(ptr, 16);
    free(ptr);
    return 0;
}
```  

2. Gen wasm
```
emcc source.c -Os -s SIDE_MODULE=1 -s EXPORTED_FUNCTIONS="['_hello_wasm']" -o target.wasm
```

3. run
```  
root@iZrj97ibqurmwnu9mj5htdZ:~/wasmer/wasmer-c-api# ./wasmer-c-api-example
Compile result:  2
Error len: `127`
Error str: `2 link errors: (1 of 2) Import not found, namespace: env, name: _free (2 of 2) Import not found, namespace: env, name: _malloc`
wasmer-c-api-example: wasmer-c-api-example.c:114: main: Assertion `compile_result == WASMER_OK' failed.
Aborted (core dumped)
```
Seem's I have to import the glibc function manually. Is there any possibility to import all glibc function automatically ?   

5. More worse
5.1 
add some standard header file into `source.c` and do nothing.
```
#include <stdio.h>
#include <malloc.h>
#include <string.h>
extern void print_str(char *ptr, int len);

int hello_wasm()
{
    char *str = "Hello, World!";
    print_str(str, 13);
    return 0;
}
```  

5.2 Compile without `-Os` of emcc flag
```
emcc wasm-sample-app/source.c  -s SIDE_MODULE=1 -s EXPORTED_FUNCTIONS="['_hello_wasm']" -o wasm-sample-app/target.wasm
``` 

5.3 Run
```
root@iZrj97ibqurmwnu9mj5htdZ:~/wasmer/wasmer-c-api# ./wasmer-c-api-example
Compile result:  2
Error len: `444`
Error str: `7 link errors: (1 of 7) Import not found, namespace: env, name: abortStackOverflow (2 of 7) Import not found, namespace: env, name: table (3 of 7) Import not found, namespace: env, name: __table_base (4 of 7) Import not found, namespace: env, name: tempDoublePtr (5 of 7) Import not found, namespace: env, name: DYNAMICTOP_PTR (6 of 7) Import not found, namespace: global, name: NaN (7 of 7) Import not found, namespace: global, name: Infinity`
wasmer-c-api-example: wasmer-c-api-example.c:114: main: Assertion `compile_result == WASMER_OK' failed.
Aborted (core dumped)
```  
I even don't know what those 7 imports are(table, DYNAMICTOP_PTR,Infinity NaN...).

**Comment by filips123** [comment]
I have a similar error with `python-ext-wasm` and similar source program.

**Comment by jponc** [comment]
You guys found any solution to this? I encountered the same thing when I try to execute a wasm binary compiled using assemblyscript

**Comment by MarkMcCaskey** [comment]
Sorry, we seemed to have missed this issue!

Our docs touch on this a bit (I just updated them to clarify a bit more): https://docs.wasmer.io/introduction

I'll also explain here:  Emscripten uses the Emscripten ABI, an ABI in the context of Wasm is an expectation of the existence of certain imported functions, memories, globals, and tables as well as things like the host/guest divide in responsibility for things like memory allocation.

When compiling with `emcc`, the host must fulfill its end of the Emscripten ABI.  Wasmer includes support for Emscripten, but it's important to note that Wasmer's Emscripten implementation is not sandboxed; it forwards system calls directly to the operating system.

Most compilers will let you control which ABI you're compiling for, `wasm32-unknown-unknown` is like an empty ABI where you can do whatever you want.  There's also `wasm32-wasi` if you want to use the WASI model, which Wasmer also includes.

@jponc it looks like AssemblyScript includes extra imports all the time for some reason which is a bit strange, you can find an explanation of how to make this work with standard WASI implementations here: https://spectrum.chat/wasmer/runtime/trying-to-run-assemblyscript-quick-start-in-wasmer~2d811551-b9b8-4c58-bdb0-a565917ad767?m=MTU3OTMwNjk5MTc5OA==

tl;dr is that these are specific to the flavor of Wasm being generated and if you're interacting with it manually then you'll have to read their documentation and figure out what they expect!

I'm going to close this issue, but feel free to respond or open a new one if you have other issues
